Classes/ios/Scanners/MTBBarcodeScanner.m

```diff
@@ -561,57 +566,60 @@ - (void)setDeviceInput:(AVCaptureDeviceInput *)deviceInput session:(AVCaptureSes
-        }
-        
-        // Focus on the center of the image
-        if ([deviceInput.device respondsToSelector:@selector(isFocusPointOfInterestSupported)] &&
-            deviceInput.device.isFocusPointOfInterestSupported) {
-            
-            self.initialFocusPoint = deviceInput.device.focusPointOfInterest;
-            deviceInput.device.focusPointOfInterest = CGPointMake(kFocalPointOfInterestX, kFocalPointOfInterestY);
-        }
-        
-        [self updateTorchModeForCurrentSettings];
-        
-        [deviceInput.device unlockForConfiguration];
-    }
-    
+    [self updateFocusPreferencesOfInput:deviceInput reset:NO];
+
     [session addInput:deviceInput];
 }
 
 - (void)removeDeviceInput {
     AVCaptureDeviceInput *deviceInput = self.currentCaptureDeviceInput;
+
     if (deviceInput == nil) {
         // No need to remove the device input if it was never set
         return;
     }
     
     // Restore focus settings to the previously saved state
-    if ([deviceInput.device lockForConfiguration:nil] == YES) {
-        if ([deviceInput.device respondsToSelector:@selector(isAutoFocusRangeRestrictionSupported)] &&
-            deviceInput.device.isAutoFocusRangeRestrictionSupported) {
-            deviceInput.device.autoFocusRangeRestriction = self.initialAutoFocusRangeRestriction;
+    [self updateFocusPreferencesOfInput:deviceInput reset:YES];
+    
+    [self.session removeInput:deviceInput];
+    self.currentCaptureDeviceInput = nil;
+}
+
+- (void)updateFocusPreferencesOfInput:(AVCaptureDeviceInput *)deviceInput reset:(BOOL)reset {
+    NSParameterAssert(deviceInput);
+
+    if (!deviceInput) {
+        return;
+    }
+
+    NSError *lockError;
+
+    if ([deviceInput.device lockForConfiguration:&lockError]) {
+        // Prioritize the focus on objects near to the device
+        if (deviceInput.device.isAutoFocusRangeRestrictionSupported) {
+            if (!reset) {
+                self.initialAutoFocusRangeRestriction = deviceInput.device.autoFocusRangeRestriction;
+            }
+
+            deviceInput.device.autoFocusRangeRestriction = reset ? self.initialAutoFocusRangeRestriction : self.preferredAutoFocusRangeRestriction;
         }
-        
-        if ([deviceInput.device respondsToSelector:@selector(isFocusPointOfInterestSupported)] &&
-            deviceInput.device.isFocusPointOfInterestSupported) {
-            deviceInput.device.focusPointOfInterest = self.initialFocusPoint;
+
+        // Focus on the center of the image
+        if (deviceInput.device.isFocusPointOfInterestSupported) {
+            if (!reset) {
+                self.initialFocusPoint = deviceInput.device.focusPointOfInterest;
+            }
+
+            deviceInput.device.focusPointOfInterest = reset ? self.initialFocusPoint : CGPointMake(kFocalPointOfInterestX, kFocalPointOfInterestY);
         }
+
+        [self updateTorchModeForCurrentSettings];
         
         [deviceInput.device unlockForConfiguration];
```

Comment:
unlock before `updateTorchModeForCurrentSettings` given that it will acquire its own lock?